design/design-sketch-oam.md

```diff
@@ -0,0 +1,334 @@
+spec:
+  # This label selector is used to limit where this application's components
+  # may run. What exactly it selects (a Kubernetes cluster, an instance
+  # group, a serverless platform, etc) is up to the runtime.
+  selector:
+    matchLabels:
+      environment: production
+      region: europe
+---
+kind: ApplicationConfiguration
+metadata:
+  namespace: default
+  name: coolest-app
+  annotations:
+    description: A very cool cloud-native application.
+spec:
+  scopes:
+  - scopeRef:
+      apiVersion: core.oam.dev/v1alpha1
+      kind: SchedulingScope
+      name: production-europe
+  components:
+  - component:
+      apiVersion: core.oam.dev/v1alpha1
+      kind: ContainerizedWorkload
+      metadata:
+        annotations:
+          version: v1.0.0
+      spec:
+        # Under this proposal the workloadType would be limited to the core OAM
+        # types, i.e. it would define the function of the resulting workload -
+        # Server, Singleton, Worker, etc. Extended workloads are a distinct kind
+        # and thus kind: ContainerizedWorkload has no workloadSettings field.
+        workloadType: core.oam.dev/v1alpha1.Server
+        osType: linux
+        containers:
+        - name: my-twitter-bot-frontend
+          image:
+            name: example/my-twitter-bot-frontend:v1.0.0
+          resources:
+            cpu:
+              required: 1.0
+            memory:
+              required: 100MB
+          ports:
+          - name: http
+            value: 8080
+          env:
+          - name: USERNAME
+            value: cooladmin
+          - name: PASSWORD
+            valueFrom:
+              secretKeyRef:
+                name: cool-frontend
+                key: password
+          - name: BACKEND_ADDRESS
+            value: https://example.org
+          livenessProbe:
+            httpGet:
+              port: 8080
+              path: /healthz
+          readinessProbe:
+            httpGet:
+              port: 8080
+              path: /healthz
+    traits:
+    # Contemporary traits model the workload kinds to which they may apply. This
+    # could be modelled as an annotation on the CRD of each Trait kind.
+    - apiVersion: core.oam.dev/v1alpha1
+      kind: ManualScalerTrait
+      spec:
+        replicaCount: 3
+  - component:
+      apiVersion: cache.crossplane.io/v1alpha1
+      kind: RedisCluster
+      spec:
+        classSelector:
+          matchLabels:
+            environment: production
+            region: europe
+        writeConnectionSecretToRef:
+          name: coolest-app-redis
+        engineVersion: "3.2"
+```
+
+Submitting the above `ApplicationController` would cause an instance of each of
+the scopes, components, and traits it templates to be created verbatim (modulo
+some generated object metadata) similar to how the `template` of a Kubernetes
+`ReplicaSet` becomes a set of `Pod` resources. For example:
+
+```yaml
+---
+apiVersion: core.oam.dev/v1alpha1
+kind: ContainerizedWorkload
```

Comment:
Actually this would be kind `ComponentSchematic`:

```yaml
kind: Component
metadata:
  name: containerized-workload
spec:
  parameters:
    ... the schema of the spec of ContainerizedWorkload ...
```

A deployment would be done in ApplicationConfiguration components section:

```yaml
kind: AppConfig
metadata: ...
spec:
  components:
  - componentName: ContainerizedWorkload
    instanceName: coolest-app-f92dm
    parameterValues: ...
```

Note that we are trying to migrate all of the schema to WorkloadType. So ideally ContainerizedWorkload is a Workload type, as the name suggested itself :)

Reply:
Similar to my comment above regarding application scopes, we're proposing a change to how OAM components are defined here. Specifically we're exploring the idea the one `ComponentSchematic` kind be broken down into many kinds of component, e.g. `ContainerizedWorkload`, `RedisCluster`, etc. The schema for each of these kinds would be defined by their custom resource definitions - not a workload type. In general we feel that Kubernetes CRDs should define configuration schema, and CRs should instantiate configuration. What do you think about this idea?